src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,133 @@
+        Book book = getBookById(requestDto.getBookId());
+
+        CartItem existingCartItem = findCartItemInCart(cart, book);
+
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(existingCartItem.getQuantity() + requestDto.getQuantity());
+        } else {
+            addNewCartItemToCart(cart, book, requestDto.getQuantity());
+        }
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    private ShoppingCart getOrCreateUserShoppingCart(Long userId) {
+        return shoppingCartRepository.findByUserId(userId)
+                .orElseGet(() -> {
+                    User user = getUserById(userId);
+                    ShoppingCart newCart = new ShoppingCart();
+                    newCart.setUser(user);
+                    return shoppingCartRepository.save(newCart);
+                });
+    }
+
+    private User getUserById(Long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("User not found with id: " + userId));
+    }
+
+    private Book getBookById(Long bookId) {
+        return bookRepository.findById(bookId)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Book not found with id: " + bookId));
+    }
+
+    private CartItem findCartItemInCart(ShoppingCart cart, Book book) {
+        return cart.getCartItems().stream()
+                .filter(cartItem -> cartItem.getBook().getId().equals(book.getId()))
+                .findFirst()
+                .orElse(null);
+    }
+
+    private void addNewCartItemToCart(ShoppingCart cart, Book book, int quantity) {
+        CartItem newCartItem = new CartItem();
+        newCartItem.setShoppingCart(cart);
+        newCartItem.setBook(book);
+        newCartItem.setQuantity(quantity);
+        cart.getCartItems().add(newCartItem);
+    }
+
+    @Override
+    public ShoppingCartDto getCartByUserId(Long userId) {
+        Optional<ShoppingCart> cart = shoppingCartRepository.findByUserId(userId);
+
+        if (cart.isPresent()) {
+            ShoppingCart shoppingCart = cart.get();
+            List<CartItemDto> cartItemDtos = shoppingCart.getCartItems().stream()
+                    .map(this::mapCartItemToDtoWithBookDetails)
+                    .collect(Collectors.toList());
+            ShoppingCartDto shoppingCartDto = shoppingCartMapper.toDto(shoppingCart);
+            shoppingCartDto.setCartItems(cartItemDtos);
+
+            return shoppingCartDto;
+        } else {
+            return null;
+        }
+    }
+
+    private CartItemDto mapCartItemToDtoWithBookDetails(CartItem cartItem) {
+        CartItemDto dto = cartItemMapper.toDto(cartItem);
+        if (cartItem.getBook() != null) {
+            dto.setBookId(cartItem.getBook().getId());
+            dto.setBookTitle(cartItem.getBook().getTitle());
+        }
+        return dto;
+    }
+
+    @Override

```

Comment:
```suggestion
    @Override
    @Transactional
```
Please don't forget about this annotation in all methods that modify something in Database and consist of more than 1 step